[프로필 화면 - 친구 탭 페이지] 친구 화면에서 친구 검색, 친구 추가, 현재 친구목록 표시, 친구 삭제 기능 구현

jungkeunking opened the pull request:
- [x] 프로필 화면에서 탭레이아웃과 뷰페이저 연동
-- 댓글, 태그 탭도 임시로 친구 화면 프래그먼트로 연동
- [x] 친구 검색 기능 구현
-- 레트로핏으로 파이어베이스 쿼리해서 시작 문자열 통해 결과 반환 (limit 기본값 10으로 설정, 반환 데이터 없을 경우 역직렬화 오류로 인해 빈값 반환하도록 예외 처리, 반환 결과에서 이미 친구거나 본인은 필터링, 반환 결과를 email 알파벳 순 정렬해서 표시, 검색 결과 없을 경우 메시지 표시)
- [x] 친구 추가 기능 구현
-- 원격 DB에서 현재 이용자의 friends 데이터 업데이트 및 UI에 실시간 반영
- [x] 현재 친구목록 표시 기능 구현
-- 파이어베이스의 현재 이용자의 friends 데이터를 flow로 변환해, 값 변경시 uiState에 실시간 반영 (레트로핏 통한 방식으로는 flow 변환이 불가능해 보여서 데이터스냅샷 활용)
- [x] 친구 삭제 기능 구현
-- 친구 추가 로직과 유사
 
- ToDo?: UserEntity에 friends 프로퍼티 추가해서 원격과 동시에 관리 (그럼 로그인시 로컬 유저 데이터도 원격과 싱크 필요)
- ToDo?: 파이어베이스 users 키값을 네이버 사용자의 경우만 자동생성 id 말고 email로 변경 (가능할진 모르겠으나 친구검색시 조회 성능 개선 가능)

피드백 반영:
- [x] 친구 추가 및 삭제 버튼 중복 클릭 방지 (친구 추가시 uiState에 친구 중복 추가 방지)
- [x] 친구 화면 파이어베이스 URL 숨김 처리, 함수명 변경, 스트링 리소스 추출

추가 수정사항:
- [x] 친구 추가/삭제 완료 후 스낵바 표시
- [x] 힐트 적용 브랜치 리베이스 및 힐트 추가 적용

추가 피드백 반영:
- [x] 비회원일 경우, 검색창 숨김 처리 및 메시지 표시

jungkeunking committed:
add: 친구추가 아이콘 추가

jungkeunking committed:
feat: 친구 화면 레이아웃 구현 중 (친구 검색, 검색 결과, 내 친구 등)

jungkeunking committed:
add: 친구 추가, 삭제 이미지 추가

jungkeunking committed:
feat: 친구 화면 및 친구, 친구찾기 아이템 레이아웃의 뷰 사이즈 및 배치 수정

jungkeunking committed:
feat: 원격DB에서 시작 단어로 유저들 검색 반환하는 함수 추가

jungkeunking committed:
feat: 친구찾기 및 친구목록 리사이클러뷰 어답터 추가

jungkeunking committed:
feat: 친구 화면에서 친구찾기 기능 및 친구목록 표시 구현

jungkeunking committed:
feat: UserDTO에 remoteId 프로퍼티 추가 및 JSON 문자열을 네이티브 객체로 변환 함수 추가

jungkeunking committed:
refactor: 탑레벨 변수명 수정 및 코드 리팩토링

jungkeunking committed:
feat: 원격DB users 데이터를 플로우로 변환해 친구 화면 UI에 친구 추가 및 삭제 실시간 반영 구현

jungkeunking committed:
add: 친구 화면 검색 결과 없음 이미지 추가

jungkeunking committed:
feat: 친구 화면 레이아웃 수정 (검색 결과 없음 메시지 표시, 검색 결과와 친구 목록 뷰 토글)

jungkeunking committed:
fix: 원격DB User 검색에서 반환값 없는 경우 예외 처리 (deserialization 오류 해결) 및 검색 결과 없음 이미지 변경

jungkeunking committed:
feat: users 테이블 전체 조회하는 방식에서 현재 사용자 데이터만 조회해서 친구 아이디 리스트를 플로우로 관리하도록 구현 수정

jsulim commented:
고생하셨습니다. 친구 추가 잘 작동하네요 친구표시 로직 고민 많이 하시는게 느껴지네요 많이 배우고 갑니다. 

근데  ToDo?: 파이어베이스 users 키값을 네이버 사용자의 경우만 자동생성 id 말고 email로 변경 (가능할진 모르겠으나 친구검색시 조회 성능 개선 가능) 
여기서 email로 키 값을 변경하면 어떤 이점이 있나요? 이미 email로 조회는 하고 있기는 한데 조회가 더 빨라지는걸까요?

juhan2103 commented:
정근님 코드 리뷰 과정에서 발견한 문제점이 있어 리뷰 남깁니다. 친구를 추가했을 때 추가한 친구가 중복되는 오류가 있는 것 같은데 수정부탁드립니다!

<img src="https://github.com/Sesac-team-03/TreasureHunt/assets/83435602/d8ce89cf-b648-4374-86c8-cd0c0736c318" width="300" height="500"/>

donghyun81 reviewed (commented):
<img width="173" alt="스크린샷 2024-02-07 오후 3 03 18" src="https://github.com/Sesac-team-03/TreasureHunt/assets/82762769/d91339e3-48ae-47b4-9fe7-c09ebd855f27">
<img width="184" alt="스크린샷 2024-02-07 오후 3 03 34" src="https://github.com/Sesac-team-03/TreasureHunt/assets/82762769/f09bb50a-360f-42a3-b6a0-95925f4f61d2">

짧은 시간내  많은 구현하시느라 고생 많으셨습니다!
혹시 me로 검색하고나서 친구 추가 하거나 검색을 취소해도 me라는 검색 내역이 남아있는데 의도하신건가요?

juhan2103 commented on app/src/main/res/layout/fragment_friend.xml:
```diff
@@ -9,31 +9,38 @@
             android:id="@+id/tiet_search_friend"
+            android:layout_width="250dp"
+            android:layout_height="40dp"
             android:hint="이메일로 친구를 검색해 보세요"
```
하드코딩 된 부분은 string 파일에서 관리하는것이 좋을 것 같습니다!

jungkeunking replied:
반영하겠습니다!

juhan2103 commented on app/src/main/java/com/treasurehunt/data/UserRepository.kt:
```diff
@@ -19,4 +19,6 @@ interface UserRepository {
 
     suspend fun delete(vararg users: UserEntity): Int
+
+    suspend fun search(startAt: String, limit: Int = 10): Map<String, UserDTO>
```
최대 10명까지만 검색되도록 구현하신 이유가 있을까요?

jungkeunking replied:
일단 제거 친구검색 참고한 레퍼런스는 카톡인데, 카톡 아이디 검색의 경우 특이하게도 정확히 일치하는 값만 결과를 반환해줍니다
저는 그건 과하다 싶어 열개로 해줬는데, 이메일 주소 특성상 해당 문자열로 시작하는 모든 이메일 계정을 조회해서 보여줄 필요가 없다고 판단했습니다
의견 있으시면 참고하겠습니다!

juhan2103 replied:
정근님이 구현하신 방법도 괜찮은 것 같습니다!

juhan2103 commented on app/src/main/java/com/treasurehunt/data/remote/UserRemoteDataSource.kt:
```diff
@@ -13,4 +14,21 @@ class UserRemoteDataSource(private val userService: UserService) {
+        Json {
+            ignoreUnknownKeys = true
+            coerceInputValues = true
+        }.decodeFromString<R>(this)
```
inline 함수 적용하신거 좋은 것 같습니다!

juhan2103 commented on app/src/main/java/com/treasurehunt/ui/profile/FriendViewModel.kt:
```diff
@@ -0,0 +1,72 @@
+    }.await()
+
+    // limit = 5
+    suspend fun search(startAt: String): Map<String, UserDTO> {
```
다른 함수들은 addFriend, removeFriend여서 search도 searchFriend로 통일하는 것이 어떨까요? 그냥 하나의 의견입니다!

jungkeunking replied:
저도 일관성 중시해서 좋은 의견입니다!
이건 친구내에서 검색하기보다 사용자내 검색이므로 searchUser 하겠습니다~

juhan2103 commented on app/src/main/java/com/treasurehunt/ui/profile/FriendViewModel.kt:
```diff
@@ -20,14 +31,95 @@ class FriendViewModel(
+    private val _uiState = MutableStateFlow(FriendUiState(false))
+    val uiState: StateFlow<FriendUiState> = _uiState
+    val db =
+        FirebaseDatabase.getInstance("https://treasurehunt-32565-default-rtdb.asia-southeast1.firebasedatabase.app")
```
이부분 같은 경우에는 UserService에서 사용했던것 처럼
```kotlin
private const val BASE_URL = BuildConfig.BASE_URL
```
이렇게 쓰는 것이 좋아보입니다. 왜냐하면 Firebase 규칙을 아마 링크가 있는 사람은 수정이 가능해서 숨겨서 관리해야 다른 사람이 건드리지 못할 것 같습니다. 규칙은 추후에 수정을 해야할 것 같긴 합니다.

jungkeunking replied:
아 중요한거 집어주셨네요 반영하겠습니다!

juhan2103 commented on app/src/main/java/com/treasurehunt/ui/profile/FriendViewModel.kt:
```diff
@@ -32,15 +43,39 @@ class FriendViewModel(
+        val uid = Firebase.auth.currentUser!!.uid
+        val friendsRef = db.reference.child("users").child(uid).child("friends")
+        val callback = object : ValueEventListener {
+            @Suppress("UNCHECKED_CAST")
```
이 어노테이션은 써도 되는 어노테이션인가요?

jungkeunking replied:
Unchecked cast suggestion을 제대로 찾아보진 않았는데 컴파일러가 안전한 타입 캐스팅인지 검증해주지 못해도 개발자가 안전성이 확실하면 써도 될거 같습니다

juhan2103 reviewed (commented):
@jungkeunking 정근님 구현 수고하셨습니다! 리뷰 남겼으니 확인 한번만 부탁드립니다!

jungkeunking commented:
> 고생하셨습니다. 친구 추가 잘 작동하네요 친구표시 로직 고민 많이 하시는게 느껴지네요 많이 배우고 갑니다.
> 
> 근데 ToDo?: 파이어베이스 users 키값을 네이버 사용자의 경우만 자동생성 id 말고 email로 변경 (가능할진 모르겠으나 친구검색시 조회 성능 개선 가능) 여기서 email로 키 값을 변경하면 어떤 이점이 있나요? 이미 email로 조회는 하고 있기는 한데 조회가 더 빨라지는걸까요?

네 좋은 질문입니다
제가 파이어베이스 쿼리 메커니즘을 제대로 이해한 건 아닌데, 현재는 유저 테이블이 아이디-객체 쌍으로 이루어져 있어서 이메일을 확인하려면 하나하나 객체를 조회해야 합니다
이메일이 아이디를 대체하면 바로 조회돼서 빠르지 않을까 했어요

jungkeunking commented:
> <img alt="스크린샷 2024-02-07 오후 3 03 18" width="173" src="https://private-user-images.githubusercontent.com/82762769/302877517-d91339e3-48ae-47b4-9fe7-c09ebd855f27.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyOTMzMDcsIm5iZiI6MTcwNzI5MzAwNywicGF0aCI6Ii84Mjc2Mjc2OS8zMDI4Nzc1MTctZDkxMzM5ZTMtNDhhZS00N2I0LTlmZTctYzA5ZWJkODU1ZjI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA3VDA4MDMyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljMWVlZDA5NDU0OTA0N2U0NTQ3MWE5NmQ5OGY2MTYwYWMxMjc3YjU3NzBhOGQzNTQwMGVmN2Q4ODUyNWZlOTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Xau9ll2v91bRZaX4J_bRD8LxK43eRxnlVgJNCvs3dXQ"> <img alt="스크린샷 2024-02-07 오후 3 03 34" width="184" src="https://private-user-images.githubusercontent.com/82762769/302877579-f09bb50a-360f-42a3-b6a0-95925f4f61d2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDcyOTMzMDcsIm5iZiI6MTcwNzI5MzAwNywicGF0aCI6Ii84Mjc2Mjc2OS8zMDI4Nzc1NzktZjA5YmI1MGEtMzYwZi00MmEzLWI2YTAtOTU5MjVmNGY2MWQyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjA3VDA4MDMyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYwZWE2NDQyN2E1YWI2OTI5YzE1ZDIxZDIzMTU4ZTM0N2U0OWI3Mzg5ZTNmNzkwY2FmMTc4OTliZmY5YzM0MmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.UcvCnntGCmQl5k4r-cYCbKYM7ozriUgFJlB6ZgfhVAk">
> 짧은 시간내 많은 구현하시느라 고생 많으셨습니다! 혹시 me로 검색하고나서 친구 추가 하거나 검색을 취소해도 me라는 검색 내역이 남아있는데 의도하신건가요?

좋은 질문입니다!
별 생각없이 했는데 하고 보니 남아 있어서 안 그래도 지우는 게 나을까 생각중이었습니다
검색 결과를 닫는 구현과 일관되려면 지우는 게 낫겠네요 반영하겠습니다

jungkeunking committed:
fix: 프로필 친구 화면에서 친구 추가 및 삭제 버튼 중복 클릭 방지 및 uiState에 친구 중복 추가 방지 및 추가/삭제 완료 후 스낵바 표시

1) 버튼 클릭리스너에 flag를 통한 조건문 활용
2) uiState friends 타입을 List -> Map 변경
3) 뷰모델의 추가/삭제 함수를 suspend로 변경후 스낵바 로직과 함께 같은 코루틴스코프에서 처리

jungkeunking commented:
> 정근님 코드 리뷰 과정에서 발견한 문제점이 있어 리뷰 남깁니다. 친구를 추가했을 때 추가한 친구가 중복되는 오류가 있는 것 같은데 수정부탁드립니다!

감사합니다! 해결했습니다

jungkeunking committed:
refactor: FriendViewModel 힐트 적용 및 프래그먼트에 반영

jungkeunking committed:
refactor: 친구 화면 파이어베이스 URL 숨김 처리, 함수명 변경, 스트링 리소스 추출

jsulim commented:
프로필 화면 관련은 아닌데 코드리뷰하면서 마커 관련 오류 발생해서 말씀드립니다. 
열린 보물상자 등록된 상태에서 열린 보물상자 밑에 부분을 클릭하면 똑같은 장소를 클릭할 수 있고 다이얼로그가 뜨게 되네요!
![image](https://github.com/Sesac-team-03/TreasureHunt/assets/144574943/deef056f-7514-48f9-8842-22e4f0e41432)

jungkeunking commented:
> 프로필 화면 관련은 아닌데 코드리뷰하면서 마커 관련 오류 발생해서 말씀드립니다. 열린 보물상자 등록된 상태에서 열린 보물상자 밑에 부분을 클릭하면 똑같은 장소를 클릭할 수 있고 다이얼로그가 뜨게 되네요! ![image](https://private-user-images.githubusercontent.com/144574943/303232273-deef056f-7514-48f9-8842-22e4f0e41432.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDczNzUyMTQsIm5iZiI6MTcwNzM3NDkxNCwicGF0aCI6Ii8xNDQ1NzQ5NDMvMzAzMjMyMjczLWRlZWYwNTZmLTc1MTQtNDhmOS04ODQyLTIyZTRmMGU0MTQzMi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIwOFQwNjQ4MzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMmY3ODUxODZiZDMxNzgyNjBkMzY4OTJhMDE0ODE2NTk5YThlNjcxNWIyNGQxZTEzYmE4Y2FmNTIzM2I4ZWJlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.5VmfkrDVsT_fgtZqX3IblfVZ5XZ0kkdIqTlnwaGrZnU)

감사합니다
디벨롭 브랜치 가서 직접 확인했습니다
새로 브랜치 파서 픽스하겠습니다~

donghyun81 commented:
익명에서는 검색이 안되는거 같은데 익명일 경우 UI를 안보여주거나 다른 화면을 보여주는게 좋지 않을까요??
그리고 현재 댓글,태그 목록에서도 친구 추가 화면이랑 같게 뜨는데 의도하신걸까요??

jungkeunking committed:
feat: 비회원 로그인 경우, 프로필 친구 화면에 메시지 표시 및 변수/함수명에서 회원을 registeredUser -> member로 수정

jungkeunking commented:
> 익명에서는 검색이 안되는거 같은데 익명일 경우 UI를 안보여주거나 다른 화면을 보여주는게 좋지 않을까요?? 그리고 현재 댓글,태그 목록에서도 친구 추가 화면이랑 같게 뜨는데 의도하신걸까요??

전자는 피드백 반영했습니다!
후자의 경우, PR 내용 첫번째로 나와 있습니다
